### PR TITLE
ORLP-424: add image and update

### PR DESCRIPTION
**Pull request by Anblak**

### CODE REVIEW

File: src/main/java/com/softserve/academy/spaced/repetition/service/DeckService.java [hunk 1]
```diff
@@ -180,4 +180,6 @@ Deck updateOwnDeck(Deck updatedDeck, Long deckId, Long categoryId)
      * @return list of cards that are in the deck by given identifier.
      */
     List<Card> getAllCardsByDeckId(Long deckId);
+
```
Comment (MoGUREV) [inline src/main/java/com/softserve/academy/spaced/repetition/service/DeckService.java]: Why did you add empty lines?

File: src/main/java/com/softserve/academy/spaced/repetition/service/cardLoaders/impl/DataSaver.java [hunk 2]
```diff
@@ -25,7 +25,7 @@ public void save(Map <String, String> map, Long deckId) {
                 Card card = new Card();
                 card.setAnswer(elem.getKey());
                 card.setQuestion(elem.getValue());
-                cardService.addCard(card, deckId);
+                cardService.addCard(card, deckId,null);
```
Comment (MoGUREV) [inline src/main/java/com/softserve/academy/spaced/repetition/service/cardLoaders/impl/DataSaver.java]: Maybe add whitespace before `null`)

File: src/main/java/com/softserve/academy/spaced/repetition/service/impl/CardImageServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+public class CardImageServiceImpl implements CardImageService {
+
+    private static final String BASE64_TYPE = "base64";
+    private static final int MIN_IMAGE_LIST_SIZE = 2;
+
+    @Autowired
+    private CardImageRepository cardImageRepository;
+
+    @Override
+    public void save(List<String> imageList, Card card) {
```
Comment (MoGUREV) [inline src/main/java/com/softserve/academy/spaced/repetition/service/impl/CardImageServiceImpl.java]: Maybe rename to AddImage or smt like that

File: src/main/java/com/softserve/academy/spaced/repetition/service/impl/CardImageServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+            if (imageList.get(0).endsWith(BASE64_TYPE)) {
+                cardImageRepository.save(new CardImage(imageList.get(0) + "," + imageList.get(1), card));
+            } else {
+                imageList.forEach(image -> cardImageRepository.save(new CardImage(image, card)));
+            }
+        }
+    }
+
+    @Override
+    public void delete(Long cardImageID) {
```
Comment (MoGUREV) [inline src/main/java/com/softserve/academy/spaced/repetition/service/impl/CardImageServiceImpl.java]: Maybe rename to deleteImageById

File: src/test/java/com/softserve/academy/spaced/repetition/util/DomainFactory.java [hunk 5]
```diff
@@ -46,7 +46,7 @@ public static Authority createAuthority(Long authorityId, AuthorityName name, Li
         authority.setAccounts(accounts);
         return authority;
     }
-  
+
```
Comment (MoGUREV) [inline src/test/java/com/softserve/academy/spaced/repetition/util/DomainFactory.java]: Why did you delete and add empty line?)

File: src/main/java/com/softserve/academy/spaced/repetition/controller/CardController.java [hunk 6]
```diff
@@ -133,41 +139,45 @@
...
+        Link selfLink = linkTo(methodOn(CardController.class).getCardByCategoryAndDeck(categoryId, deckId, card.getId())).withSelfRel();
+
         CardPublicDTO cardPublicDTO = DTOBuilder.buildDtoForEntity(card, CardPublicDTO.class, selfLink);
         return new ResponseEntity<>(cardPublicDTO, HttpStatus.OK);
     }
 
     @Auditable(action = AuditingAction.EDIT_CARD_VIA_CATEGORY_AND_DECK)
-    @PutMapping(value = "/api/decks/{deckId}/cards/{cardId}")
+    @PostMapping(value = "/api/decks/{deckId}/cards/{cardId}")
     @PreAuthorize(value = "@accessToUrlService.hasAccessToCard(#deckId, #cardId)")
```
Comment (VovaChornyi) [inline src/main/java/com/softserve/academy/spaced/repetition/controller/CardController.java]: You must use PUT Mapping by REST API convension

File: src/main/java/com/softserve/academy/spaced/repetition/service/DeckService.java (hunk 1 above)
Comment (Anblak) [reply src/main/java/com/softserve/academy/spaced/repetition/service/DeckService.java]: :З

File: src/test/java/com/softserve/academy/spaced/repetition/service/CardServiceTest.java [hunk 7]
```diff
@@ -268,31 +301,19 @@ public void testUploadCards() throws WrongFormatException, EmptyFileException, N
...
-        verify(cardsFile).getContentType();
         verify(cardsFile).isEmpty();
         verify(cardsFile).getInputStream();
         verify(deckRepository).findOne(DECK_ID);
 
     }
 
-    @Test(expected = WrongFormatException.class)
-    public void testUploadCardsWrongFormat() throws WrongFormatException, EmptyFileException, NotOwnerOperationException,
-            NotAuthorisedUserException, IOException {
```
Comment (MoGUREV) [inline src/test/java/com/softserve/academy/spaced/repetition/service/CardServiceTest.java]: Why did you delete this test?

File: src/main/java/com/softserve/academy/spaced/repetition/controller/CardController.java (hunk 6 above)
Comment (Anblak) [reply src/main/java/com/softserve/academy/spaced/repetition/controller/CardController.java]: in another task

File: src/test/java/com/softserve/academy/spaced/repetition/service/CardServiceTest.java (hunk 7 above)
Comment (Anblak) [reply src/test/java/com/softserve/academy/spaced/repetition/service/CardServiceTest.java]: remove validation

File: src/main/java/com/softserve/academy/spaced/repetition/domain/CardImage.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+import javax.persistence.*;
+import javax.validation.constraints.NotNull;
+
+@Entity
+@Table(name = "card_image")
+public class CardImage implements EntityInterface {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    @Column(name = "card_image_id", nullable = false)
```
Comment (serhiivenhrynovych) [inline src/main/java/com/softserve/academy/spaced/repetition/domain/CardImage.java]: you don`t need to use nullable = false in ID
remove it

File: src/main/java/com/softserve/academy/spaced/repetition/domain/CardImage.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    private Long id;
+
+    @Column(name = "imagebase64", columnDefinition = "LONGTEXT")
```
Comment (serhiivenhrynovych) [inline src/main/java/com/softserve/academy/spaced/repetition/domain/CardImage.java]: change name of column to image_base64

File: src/main/java/com/softserve/academy/spaced/repetition/controller/CardController.java [hunk 10]
```diff
@@ -97,31 +99,34 @@
...
+        Link selfLink = linkTo(methodOn(CardController.class).getCardByCourseAndDeck(categoryId, courseId, deckId, card.getId())).withSelfRel();
         CardPublicDTO cardPublicDTO = DTOBuilder.buildDtoForEntity(card, CardPublicDTO.class, selfLink);
         return new ResponseEntity<>(cardPublicDTO, HttpStatus.CREATED);
     }
 
     @Auditable(action = AuditingAction.CREATE_CARD_VIA_CATEGORY_AND_DECK)
     @PostMapping(value = "/api/category/{categoryId}/decks/{deckId}/cards")
     @PreAuthorize(value = "@accessToUrlService.hasAccessToDeckFromCategory(#categoryId, #deckId)")
-    public ResponseEntity<CardPublicDTO> addCardByCategoryAndDeck(@Validated(Request.class) @RequestBody Card card,
+    public ResponseEntity<CardPublicDTO> addCardByCategoryAndDeck(@RequestParam String title,
```
Comment (VovaChornyi) [inline src/main/java/com/softserve/academy/spaced/repetition/controller/CardController.java]: use @RequstBody instead all requestParams

Comment (Anblak) [reply src/main/java/com/softserve/academy/spaced/repetition/controller/CardController.java]: there are many problems with it

Comment (RomanSomyk) [reply src/main/java/com/softserve/academy/spaced/repetition/controller/CardController.java]: That's why you are learning it, if you solve these problems it will be kind of improvement of your skills

File: src/main/java/com/softserve/academy/spaced/repetition/service/CardImageService.java [hunk 11]
```diff
@@ -0,0 +1,13 @@
+package com.softserve.academy.spaced.repetition.service;
+
+import com.softserve.academy.spaced.repetition.domain.Card;
+import com.softserve.academy.spaced.repetition.domain.CardImage;
+
+import java.util.List;
+
+public interface CardImageService {
+
```
Comment (serhiivenhrynovych) [inline src/main/java/com/softserve/academy/spaced/repetition/service/CardImageService.java]: add javadocs
